README.md
## Dataspace Connector UI

User interface for the [Dataspace Connector](https://github.com/International-Data-Spaces-Association/DataspaceConnector).

The following illustration visualizes the interaction of the [Dataspace Connector](https://github.com/International-Data-Spaces-Association/DataspaceConnector), the [IDS Messaging Services](https://github.com/International-Data-Spaces-Association/IDS-Messaging-Services), the Configuration Manager, and it’s GUI. All components have a defined API that allows individual components to be removed or replaced. The Dataspace Connector can be deployed standalone and can be connected to existing backend systems. Configuration Manager and GUI facilitate the operation and configuration of the connector. If desired, the Dataspace Connector may be replaced by another connector implementation, either integrating the IDS Messaging Services or not.

![Overall architecture](https://github.com/International-Data-Spaces-Association/DataspaceConnector/blob/main/docs/assets/images/dsc_architecture.png)

## Requirements
* Install and start [Dataspace Connector](https://github.com/International-Data-Spaces-Association/DataspaceConnector)
* Install [Node.js v.14](https://nodejs.org/en/download/)

## Installation

Use the package manager [npm](https://nodejs.org/en/download/) in root directory to install DataspaceConnector UI:

```bash
npm install --no-audit
```

## Usage

### Start Dataspace Connector UI
Use the package manager [npm](https://nodejs.org/en/download/) in root directory to start Dataspace Connector UI:
```bash
npm start
```
Access: [localhost:8082](http://localhost:8082) 

### Change UI port

Change `package.json` in root directory:
```bash
"serve": "vue-cli-service serve --open --port [PORT]"
```

### Change Dataspace Connector host & port

Change in `src/backend/index.js`:
```bash
let connectorUrl = "https://localhost:8080"
```

### Change theme

You can change the main colors of the user interface in `src/theme/default.js`

### Protect backend with Basic Auth
If you need to protect the website via Basic Auth, you can adjust the username and password via environment variables of the backend, simply by adjusting the `src/backend/.env` file or by setting the environment variables in your Docker environment.
To activate the protection, you need to provide a username and a password, both with a length > 0. Otherwise (default behaviour), it is inactive.
```
BASIC_AUTH_USER=username
BASIC_AUTH_PASSWORD=password
```

### Test backend

The UI backend provides an endpoint (http://[localhost:8083]/testdata) that can be used as backend connection (type: REST) in the DSC for testing purposes.
This can process POST and GET requests.

## Start with Docker

Build docker image:
```bash
sudo docker build -t dataspace-connector-ui .
```
Run docker image:
```bash
sudo docker-compose up
```
Access: [localhost:8083](http://localhost:8083) 

### Change host/port & authentification of Dataspace Connector on docker start

Change in `docker-compose.yml`:
```bash
environment:
          - CONNECTOR_URL=https://localhost:8080
          - CONNECTOR_USER=testuser
          - CONNECTOR_PASSWORD=testpw
```

## Add custom attributes to Resources
The Dataspace Connector allows to define additional attributes for each of its entities (see [Dataspace Connector Documentation](https://international-data-spaces-association.github.io/DataspaceConnector/Documentation/v6/DataModel). If the additional fields are set for a `Resource` entity, this information can be used to provider further metadata, e.g. for registering these data at an IDS Metadata Broker (which must support these Metadata-fields too). The Dataspace Connector UI provides the functionality to define custom fields, which are shown in the frontend to define such additional information on a Resource level (see Screenshot below). By default the functionality is disabled and thus not shown in the UI.

![Screenshot of additional meta information for a Resource offering](images/feature-additional-parameters.png)

You can define additional elements via three input options: 
- text fields
- select boxes (e.g., a category)
- select boxes with according sub-select boxes (e.g., category with sub-categories) 

### Enable and adjust the configuration
To enable the feature, go to the backend folder `src/backend/` and adjust `.env` file: `USE_ONTOLOGY = true`. In case you are deploying the frontend via Docker set environment variables accordingly. The processing of the feature takes place in `ontologyLoader.js`. A demo configuration is already given by the files `src/backend/ontology.ttl` and `src/backend/ontology.config.json`.

#### Adjust the ontology
You need to adjust the content of the file `src/backend/ontology.ttl`, which already provides some examples.

##### Select boxes
For select boxes, you have to define an `owl:class` element:
```
dsc:DataCategory 
    a owl:Class ;
    rdfs:label "Data category"@en ;
    rdfs:comment "Class of all data categories."@en .
                 
dsc:dataCategory 
    a owl:ObjectProperty ;
    rdfs:domain ids:Resource ;
    rdfs:range dsc:DataCategory ;
    rdfs:label "Data category"@en .
```

If you want to define a select box with a sub-select box, in addition you need to add (do not forget to link the elements of the class to this sub-category (see below)):
```
dsc:DataSubcategory 
    a owl:Class ;
    rdfs:label "Data subcategory"@en ;
    rdfs:comment "Class of all data subcategories."@en .
                    
dsc:dataSubcategory 
    a owl:ObjectProperty ;
    rdfs:domain ids:Resource ;
    rdfs:range dsc:DataSubcategory ;
    rdfs:label "Data subcategory"@en .
```

To add categories to the root select box, simply add elements referring to the class:
```
cat:Infrastructure 
    a dsc:DataCategory ;
    rdfs:label "Infrastructure"@en .
```

To add elements for sub-select boxes, add them accordingly: 
```
sub:RoadNetwork 
    a cat:Infrastructure ;
    rdf:type   dsc:DataSubcategory ;
    rdfs:label "Road Network"@en .
```

##### Text fields
For text fields, you have to define an `owl:DatatypeProperty` element:
```
dsc:dataModel 
    a owl:DatatypeProperty ;
    rdfs:domain ids:Resource ;
    rdfs:range rdfs:Literal ;
    rdfs:label "Data model"@en .
```

#### Customizing
In case you cannot stick to the given ontology, you may change the static variables for predicates and objects to identify the structure of the given ontology in `ontologyLoader.js`. 

#### Adjust the configuration `ontology.config.json`
In order to map and display the additional fields, you need to adjust the configuration file. Each select box needs to be defined as object in `select` array, and text fields need to be defined as objects in `text` array.
```
{
  "select":[
    {...}, ..., {...}
  ],
  "text":[
    {...}, ..., {...}
  ]
}
```

#### Select box configuration
You need to define four properties on each object:
- `"identifier": string`: The identifier of the ontology class for a select box, e.g. `http://w3id.org/dsc#DataCategory`. This identifier is used to analyze the structure of your ttl file. Be aware that this identifier is overwritten within the process of parsing, as the expected key needs to be a parameter instead of a class. Thus, in this example `http://w3id.org/dsc#dataCategory` would be the key (notice the lower case `d`)
- `"required": boolean`: If it is required to select an option from the select box.
- `"identifier_children": string`: The identifier of the ontology class for a sub-select box, e.g. `http://w3id.org/dsc#DataSubcategory`. This field is not used for parsing. It is just used for the front-end to determine the correct key for the selected value.
- `"required_children": boolean`: If it is required to select an option from the sub-select box.

#### Text field configuration
You need to define two properties on each object:
- `"identifier": string`: The identifier of the ontology class for a text field, e.g. `http://w3id.org/dsc#dataModel`.
- `"required": boolean`: If it is required to enter text in the text field.


## Development
Please read the [development guide](https://github.com/International-Data-Spaces-Association/DataspaceConnectorUI/blob/develop/DEVELOPMENT_GUIDE.md).


## Contributing
Please read through our [contributing guidelines](https://github.com/International-Data-Spaces-Association/DataspaceConnectorUI/blob/develop/CONTRIBUTING.md).

### Main Contact
[Sebastian Opriel](https://github.com/SebastianOpriel), [sovity GmbH](https://sovity.de)

### Contributors
* [Bastian Wetljen](https://github.com/BastianWel), [Fraunhofer FKIE](https://www.fkie.fraunhofer.de)
* [Sebastian Opriel](https://github.com/SebastianOpriel), [sovity GmbH](https://sovity.de)
* [Melissa Das](https://github.com/melissadas), [sovity GmbH](https://sovity.de)
* [Philipp Reusch](https://github.com/pdrd), [sovity GmbH](https://sovity.de)

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.


## [10.2.2] - 2022-12-06

### Changed
- Updated Dependencies


## [10.2.1] - 2022-11-14

### Fixed
- EndpointDocumentation accessed on wrong IDS infomodel object level


## [10.2.0] - 2022-11-03

### Added
- Endpoint Documentation to Meta Data page

### Fixed
- Fixed the UI to get the active configuration of the DSC

### Changed
- renamed variable `standardlicense` to `standardLicense`


## [10.1.0] - 2022-10-06

### Fixed
- Log ontology if DEBUG is enabled
- Policy in Policy Template is doubled

### Removed
- entrypoint.sh now is a direct command in Dockerfile
- buildDockerImage.sh Instead use `sudo docker build -t dataspace-connector-ui .`

### Changed
- Renamed default title and description of UI
- Prevent adding contract templates with duplicate titles by adding "(1)" to its title


## [10.0.1] - 2022-08-19

### Added
- Connector ID adjustable via Settings page

### Fixed
- Inconsistent keyStore and trustStore variable naming between UI and DSC
- Ontology page view undefined on view of a specific resource
- Policy template checkbox not loaded correctly on view
- Button "Register data source" links to blank page
- Txt file upload with file larger than 1 MB failed. Set max file upload to 4 GB

### Removed
- Subtitle of dashboard page


## [10.0.0] - 2022-05-23 (compatible with DSC 7.1.0)
**_ATTENTION: This release requires at least DSC version 7.1.0, which introduces `addition` field in EndpointView. See [Changelog DSC v7.1.0](https://github.com/International-Data-Spaces-Association/DataspaceConnector/releases/tag/v7.1.0,changelog)_**

### Added
- Functionality to add custom properties to Resources (see [README.md](https://github.com/International-Data-Spaces-Association/DataspaceConnectorUI#add-custom-attributes-to-resources))
- Functionality to add a Basic Auth to protect the backend (see [README.md](https://github.com/International-Data-Spaces-Association/DataspaceConnectorUI#protect-backend-with-basic-auth))
- Functionality to set DeployMode of a Connector via Settings
- Functionality to manage whitelisted DAPS
- Dashboard card showing the configured default endpoint
- Dashboard card showing number of data sources
- Dashboard card showing number of policy templates
- Dashboard card showing number of data offerings
- Dashboard card showing number of active contracts
- Customizable background color of navigation via variable `navigationBackground`
- Health check in backend for kubernetes probes
- Title and description for data sources
- Templating for policies (Contract + Rules as templates)
- Advanced mode toggle (bottom of navigation bar) hiding in default's disabled mode several navigation bar elements
- Header shows connector description
- Check if a contract's validity time range fits to the current date, otherwise do not show buttons to request artifact
- Display title in backend connection remove dialog instead of URL
- Display title in backend connection select dialog

### Changed
- Several labels to explain functionalities simpler
- Navigation elements structure
- Adjusted shown fields in Add and Adjust popup for Backend Connections
- Colors of dropdown elements
- Sticky logo on small screens showing scroll bar in navigation
- Moved CSS styles from App.vue into separate default.css
- Updated package.json version number

### Fixed
- Screen refreshing when pressing enter in URL field at Data Consumption view
- Raised a broad range of comparisons of variables from value to type level
- Validation for database URLs does not require http prefix

### Removed
- Unused imports

### Dependencies
- @vue/cli-plugin-babel: 4.5.15 -> ^5.0.4
- @vue/cli-plugin-eslint: 4.5.15 -> ~5.0.0
- @vue/cli-service: ^4.5.15 -> ^5.-0.4
- eslint: ^6.7.2, -> ^7.32.0
- eslint-plugin-vue: ^6.2.2 -> ^8.0.3
- ADDED @babel/core: ^7.12.16
- ADDED @babel/eslint-parser: ^7.12.16

## [10.0.0] - 2022-07-04 (compatible with DSC 7.1.0)
**_ATTENTION: This release requires at least DSC version 7.1.0, which introduces `addition` field in EndpointView. See [Changelog DSC v7.1.0](https://github.com/International-Data-Spaces-Association/DataspaceConnector/releases/tag/v7.1.0,changelog)_**

### Added
- Functionality to add custom properties to Resources (see [README.md](https://github.com/International-Data-Spaces-Association/DataspaceConnectorUI#add-custom-attributes-to-resources))
- Functionality to add a Basic Auth to protect the backend (see [README.md](https://github.com/International-Data-Spaces-Association/DataspaceConnectorUI#protect-backend-with-basic-auth))
- Functionality to set DeployMode of a Connector via Settings
- Functionality to manage whitelisted DAPS
- Dashboard card showing the configured default endpoint
- Dashboard card showing number of data sources
- Dashboard card showing number of policy templates
- Dashboard card showing number of data offerings
- Dashboard card showing number of active contracts
- Customizable background color of navigation via variable `navigationBackground`
- Health check in backend for kubernetes probes
- Title and description for data sources
- Templating for policies (Contract + Rules as templates)
- Advanced mode toggle (bottom of navigation bar) hiding in default's disabled mode several navigation bar elements
- Header shows connector description
- Check if a contract's validity time range fits to the current date, otherwise do not show buttons to request artifact
- Display title in backend connection remove dialog instead of URL
- Display title in backend connection select dialog

### Changed
- Several labels to explain functionalities simpler
- Navigation elements structure
- Adjusted shown fields in Add and Adjust popup for Backend Connections
- Colors of dropdown elements
- Sticky logo on small screens showing scroll bar in navigation
- Moved CSS styles from App.vue into separate default.css
- Updated package.json version number

### Fixed

src/backend/index.js
import 'dotenv/config'
import express from "express";
import cors from "cors";
import axios from "axios";
import https from "https";
import bodyParser from "body-parser";
import path from "path";
import ontologyLoader from "./ontologyLoader.js"
import basicAuth from "express-basic-auth";

const vuePath = path.join(path.resolve(),'..', '..', 'dist');
const app = express();
const port = 8083;

let DEBUG = false;
let connectorUrl = "https://localhost:8080"
let auth = {
    username: 'admin',
    password: 'password'
}
let httpsAgent = new https.Agent({
    maxVersion: "TLSv1.2",
    minVersion: "TLSv1.2",
    rejectUnauthorized: false
    //ca: fs.readFileSync('dsc.crt')
});

let basicAuthUser = process.env.BASIC_AUTH_USER || '';
let basicAuthPassword = process.env.BASIC_AUTH_PASSWORD || '';

if (process.env.USE_ONTOLOGY !== undefined && process.env.USE_ONTOLOGY === "true") {
    ontologyLoader.loadOntology();
}

if (process.env.CONNECTOR_URL !== undefined) {
    connectorUrl = process.env.CONNECTOR_URL;
}

if (process.env.DEBUG !== undefined) {
    DEBUG = process.env.DEBUG === "true";
}

if (process.env.CONNECTOR_USER !== undefined) {
    auth.username = process.env.CONNECTOR_USER;
}

if (process.env.CONNECTOR_PASSWORD !== undefined) {
    auth.password = process.env.CONNECTOR_PASSWORD;
}

// initialize health before basicAuth to allow access without authentication
app.use('/health', function (req, res) {
    res.end('OK')
});

// require basicauth implicitly if user and pass are set via env to strings with length > 0
if(typeof basicAuthUser === 'string' && basicAuthUser.length > 0
    && typeof basicAuthPassword === 'string' && basicAuthPassword.length > 0) {
    console.log('Enabling basic authentication');

    let basicAuthOptions = {
        users: {},
        challenge: true
    };

    basicAuthOptions.users[basicAuthUser] = basicAuthPassword;
    app.use(basicAuth(basicAuthOptions));
}

app.use(express.static(vuePath));
app.use(cors({ credentials: true, origin: true }));
app.use(bodyParser.urlencoded({
    extended: true,
    limit: '4096mb'
}));
app.use(bodyParser.json({limit: '4096mb'}));

function post(url, data) {
    if (DEBUG === true) {
        console.log(">>> POST " + url);
        console.log(">>> DATA: ", data);
    }
    return axios.post(url, data, {
        headers: { 'content-type': 'application/json' },
        auth,
        httpsAgent
    });
}

function put(url, data) {
    if (DEBUG === true) {
        console.log(">>> PUT " + url);
        console.log(">>> DATA: ", data);
    }
    return axios.put(url, data, {
        headers: { 'content-type': 'application/json' },
        auth,
        httpsAgent
    });
}

function get(url) {
    if (DEBUG === true) {
        console.log(">>> GET " + url);
    }
    return axios.get(url, {
        headers: { 'content-type': 'application/json' },
        auth,
        httpsAgent
    });
}

function del(url, data) {
    if (DEBUG === true) {
        console.log(">>> DELETE " + url);
        console.log(">>> DATA: ", data);
    }
    return axios.delete(url, {
        data: data,
        headers: { 'content-type': 'application/json' },
        auth,
        httpsAgent
    });
}

function escape(text, key) {
    let escapedText = "";
    if (Array.isArray(text)) {
        for (let i = 0; i < text.length; i++) {
            if (i > 0) {
                escapedText += "&" + key + "=";
            }
            escapedText += encodeURIComponent(text[i]);
        }
    } else {
        escapedText = encodeURIComponent(text);
    }
    return escapedText;
}

function stringifySafe(obj, replacer, spaces, cycleReplacer) {
    return JSON.stringify(obj, serializer(replacer, cycleReplacer), spaces)
}

function filterError(origError) {
    if (origError.response !== undefined) {
        // remove config & request data, because it contains sensitive data.
        origError.response.config = null;
        origError.response.request = null;
    }
    return origError;
}

function serializer(replacer, cycleReplacer) {
    var stack = [], keys = []

    if (cycleReplacer == null) cycleReplacer = function (key, value) {
        if (stack[0] === value) return "[Circular ~]"
        return "[Circular ~." + keys.slice(0, stack.indexOf(value)).join(".") + "]"
    }

    return function (key, value) {
        if (stack.length > 0) {
            var thisPos = stack.indexOf(this)
            ~thisPos ? stack.splice(thisPos + 1) : stack.push(this)
            ~thisPos ? keys.splice(thisPos, Infinity, key) : keys.push(key)
            if (~stack.indexOf(value)) value = cycleReplacer.call(this, key, value)
        }
        else stack.push(value)

        return replacer == null ? value : replacer.call(this, key, value)
    }
}

app.get('/ontology', function (req, res) {
    res.send(ontologyLoader.getOntology());
});

app.get('/testdata', function (req, res) {
    res.send("TEST DATA FROM BACKEND");
});

app.post('/testdata', function (req, res) {
    console.log(">>> TEST DATA RECEIVED: ", req.body);
    res.send("OK");
});

app.get('/', function (req, res) {
    res.sendFile(path.join(vuePath, "index.html"));
});

app.post('/', (req, res) => {
    let call = req.body.url;
    let i = 0;
    for (let key in req.body.params) {
        if (i === 0) {
            call += "?" + key + "=" + escape(req.body.params[key], key);
        } else {
            call += "&" + key + "=" + escape(req.body.params[key], key);
        }
        i++;
    }
    if (req.body.type == "POST") {
        post(connectorUrl + call, req.body.body).then(response => {
            res.send(response.data);
        }).catch(origError => {
            let error = filterError(origError);
            if (error.response === undefined) {
                console.log("Error on POST " + req.body.url, error);
                res.send(stringifySafe(error));
            } else {
                console.log("Error on POST " + req.body.url, stringifySafe(error.response));
                res.send(stringifySafe(error.response));
            }

        });
    } else if (req.body.type == "PUT") {
        put(connectorUrl + call, req.body.body).then(response => {
            res.send(response.data);
        }).catch(origError => {
            let error = filterError(origError);
            if (error.response === undefined) {
                console.log("Error on PUT " + req.body.url, error);
                res.send(stringifySafe(error));
            } else {
                console.log("Error on PUT " + req.body.url, stringifySafe(error.response));
                res.send(stringifySafe(error.response));
            }
        });
    } else if (req.body.type == "GET") {
        if (call == "/connector/address") {
            res.send(connectorUrl);
        } else {
            get(connectorUrl + call).then(response => {
                res.send(response.data);
            }).catch(origError => {
                let error = filterError(origError);
                if (error.response === undefined) {
                    console.log("Error on GET " + req.body.url, error);
                    res.send(stringifySafe(error));
                } else {
                    console.log("Error on GET " + req.body.url, stringifySafe(error.response));
                    res.send(stringifySafe(error.response));
                }
            });
        }
    } else if (req.body.type == "DELETE") {
        del(connectorUrl + call, req.body.body).then(response => {
            res.send(response.data);
        }).catch(origError => {
            let error = filterError(origError);
            if (error.response === undefined) {
                console.log("Error on DELETE " + req.body.url, error);
                res.send(stringifySafe(error));
            } else {
                console.log("Error on DELETE " + req.body.url, stringifySafe(error.response));
                res.send(stringifySafe(error.response));
            }
        });
    }
});

app.listen(port, () => {
    console.log(`Backend listening at http://localhost:${port}`)
});
